cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1620 @@
+
+        // The format of the result:
+        // a list of strings containing key, value pairs with colon as separator
+        // each is the size of a buffer pool
+
+        for ( auto i : ret)
+        {
+            auto pairs = tokenize(i, ':');
+            auto poolName = pairs[0];
+
+            if ("debug" != poolName)
+            {
+                auto &pool = m_bufferPoolLookup[pairs[0]];
+
+                if (pool.total_size == pairs[1])
+                    continue;
+
+                pool.total_size = pairs[1];
+
+                if (pool.initialized)
+                {
+                    updateBufferPoolToDb(poolName, pool);
+                }
+                else
+                {
+                    updateBufferPoolToDb(poolName, pool);
+                    pool.initialized = true;
+                }
+
+                SWSS_LOG_NOTICE("Buffer pool %s had been updated with new size [%s]", poolName.c_str(), pool.total_size.c_str());
+            }
+            else
+            {
+                SWSS_LOG_INFO("Buffer pool debug info %s", i.c_str());
+            }
+        }
+    }
+    catch (...)
+    {
+        SWSS_LOG_WARN("Lua scripts for buffer calculation were not executed successfully");
+    }
+}
+
+void BufferMgrDynamic::checkSharedBufferPoolSize()
+{
+    // PortInitDone indicates all steps of port initialization has been done
+    // Only after that does the buffer pool size update starts
+    if (!m_portInitDone)
+    {
+        vector<FieldValueTuple> values;
+        if (m_applPortTable.get("PortInitDone", values))
+        {
+            SWSS_LOG_NOTICE("Buffer pools start to be updated");
+            m_portInitDone = true;
+        }
+        else
+        {
+            if (m_firstTimeCalculateBufferPool)
+            {
+                // It's something like a placeholder especially for warm reboot flow
+                // without all buffer pools created, buffer profiles are unable to be cureated,
+                // which in turn causes buffer pgs and buffer queues unable to be created,
+                // which prevents the port from being ready and eventually fails the warm reboot
+                // After the buffer pools are created for the first time, we won't touch it
+                // until portInitDone
+                // Eventually, the correct values will pushed to APPL_DB and then ASIC_DB
+                recalculateSharedBufferPool();
+                m_firstTimeCalculateBufferPool = false;
+                SWSS_LOG_NOTICE("Buffer pool update defered because port is still under initialization, start polling timer");
+            }
+
+            return;
+        }
+    }
+
+    if (!m_mmuSize.empty())
+        recalculateSharedBufferPool();
+}
+
+// For buffer pool, only size can be updated on-the-fly
+void BufferMgrDynamic::updateBufferPoolToDb(const string &name, const buffer_pool_t &pool)
+{
+    vector<FieldValueTuple> fvVector;
+
+    if (pool.ingress)
```

Comment:
Can we keep a knob to not update buffer pool for lossy and lossless and just update the profile based on speed/cable length??? Some asic can fix the upper max lossy or lossless pool size, but don't want to change while packets are flowing.

Reply:
Yes. It can be done by providing ‘size’ for those pools.